File: src/op/conv2d.cc
#include "op/conv2d.h"

#include <tuple>

#include "utils/utils.h"
#include "tensor/tensor.cuh"
#include "backend/cuda/conv2d.cuh"

namespace TinyTorch {

Tensor conv2d_forward(const Tensor &input, const Tensor &weight, const Tensor &bias, int stride, int padding) {
    ASSERT(input.device.is_cuda() &&
           weight.device.is_cuda() &&
           bias.device.is_cuda());
    ASSERT(input.dim() == 4 &&
           weight.dim() == 4 && 
           bias.dim() == 1);
    int batchsize = input.shape[0];
    int channels_in = weight.shape[1];
    int channels_out = weight.shape[0];
    int height = input.shape[2];
    int width = input.shape[3];
    int kernel_size = weight.shape[2];
    ASSERT(weight.shape[3] == kernel_size &&
           input.shape[1] == channels_in &&
           bias.shape[0] == channels_out);
    int height_out = (height + 2 * padding - kernel_size) / stride + 1;
    int width_out = (width + 2 * padding - kernel_size) / stride + 1;
    Tensor output({batchsize, channels_out, height_out, width_out}, input.device);
    if (input.device.is_cuda()) {
        Backend::CUDA::conv2d_forward(input.data, output.data, weight.data, bias.data,
                                      batchsize, channels_in, channels_out, height, width,
                                      kernel_size, padding, stride, height_out, width_out);
    } else {}
    return output;
}

std::tuple<Tensor, Tensor, Tensor> conv2d_backward(const Tensor &input, const Tensor &weight, const Tensor &grad_output,
                       int stride, int padding) {
    ASSERT(input.device.is_cuda() &&
           weight.device.is_cuda() &&
           grad_output.device.is_cuda());
    ASSERT(input.dim() == 4 &&
           weight.dim() == 4 &&
           grad_output.dim() == 4);
    int batchsize = input.shape[0];
    int channels_in = weight.shape[1];
    int channels_out = weight.shape[0];
    int height = input.shape[2];
    int width = input.shape[3];
    int kernel_size = weight.shape[2];
    int height_out = (height + 2 * padding - kernel_size) / stride + 1;
    int width_out = (width + 2 * padding - kernel_size) / stride + 1;
    ASSERT(weight.shape[3] == kernel_size &&
           input.shape[1] == channels_in &&
           grad_output.shape[0] == batchsize &&
           grad_output.shape[1] == channels_out &&
           grad_output.shape[2] == height_out &&
           grad_output.shape[3] == width_out);
    Tensor grad_input(input.shape, input.device);
    Tensor grad_weight(weight.shape, input.device);
    Tensor grad_bias({channels_out}, input.device);
    if (input.device.is_cuda()) {
        Backend::CUDA::conv2d_backward(input.data, nullptr, weight.data, nullptr,
                                       batchsize, channels_in, channels_out, height, width,
                                       kernel_size, padding, stride, height_out, width_out,
                                       grad_output.data, grad_input.data, grad_weight.data, grad_bias.data);
    } else {}
    return std::make_tuple(std::move(grad_input), std::move(grad_weight), std::move(grad_bias));
}

}

File: unittest/conv2d.py
import tinytorch

import torch
from torch import nn
from torch.nn import functional as F

import numpy as np

BATCHSIZE = 64
CHANNELS_IN = 3
CHANNELS_OUT = 6
HEIGHT = 28
WIDTH = 28
KERNEL_SIZE = 3
STRIDE = 1
PADDING = 1

HEIGHT_OUT = (HEIGHT + 2 * PADDING - KERNEL_SIZE) // STRIDE + 1
WIDTH_OUT = (WIDTH + 2 * PADDING - KERNEL_SIZE) // STRIDE + 1

device = None

def test_conv2d():
    input = torch.randn((BATCHSIZE, CHANNELS_IN, HEIGHT, WIDTH), device=device, requires_grad=True)
    weight = torch.randn((CHANNELS_OUT, CHANNELS_IN, KERNEL_SIZE, KERNEL_SIZE), device=device, requires_grad=True)
    bias = torch.randn((CHANNELS_OUT, ), device=device, requires_grad=True)
    grad_output = torch.randn((BATCHSIZE, CHANNELS_OUT, HEIGHT_OUT, WIDTH_OUT), device=device, requires_grad=True)
    output_gt = F.conv2d(input, weight, bias, STRIDE, PADDING)
    output_gt.backward(grad_output)

    output_gt = tinytorch.Tensor(output_gt.detach().numpy())
    grad_input_gt = tinytorch.Tensor(input.grad.detach().numpy())
    grad_weight_gt = tinytorch.Tensor(weight.grad.detach().numpy())
    grad_bias_gt = tinytorch.Tensor(bias.grad.detach().numpy())
    input = tinytorch.Tensor(input.detach().numpy())
    weight = tinytorch.Tensor(weight.detach().numpy())
    bias = tinytorch.Tensor(bias.detach().numpy())
    grad_output = tinytorch.Tensor(grad_output.detach().numpy())

    output_res = tinytorch.op.conv2d_forward(input, weight, bias, STRIDE, PADDING)
    grad_input_res, grad_weight_res, grad_bias_res = tinytorch.op.conv2d_backward(input, weight, grad_output, STRIDE, PADDING)

    # if errors accumulate and assert fails, `assert tinytorch.op.tensor_close(res, gt)` may be useful,
    # which is a looser test
    assert output_res == output_gt
    assert grad_input_res == grad_input_gt
    assert grad_weight_res == grad_weight_gt
    assert grad_bias_res == grad_bias_gt

def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print("Torch Device:", device)
    tinytorch.Device.set_default_device(tinytorch.Device.cuda())
    print("TinyTorch Default Device:", "cpu" if tinytorch.Device.get_default_device().is_cpu() else "cuda")
    test_conv2d()

if __name__ == "__main__":
    main()
